DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/zero_copy_data_transfer/src/buffer_kernel.hpp

```diff
@@ -0,0 +1,55 @@
+#ifndef __BUFFER_KERNEL_HPP__
+#define __BUFFER_KERNEL_HPP__
+#pragma once
+
+#include <vector>
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/fpga_extensions.hpp>
+
+using namespace sycl;
+using namespace std::chrono;
+
+// Declare the kernel class name globally to avoid name mangling.
+class BufferWorker;
+
+// kernel using buffers to transfer data
+template <typename T>
+double BufferKernel(queue& q, std::vector<T>& in, std::vector<T>& out,
+                    const unsigned int size) {
+  // start timer
+  auto start = high_resolution_clock::now();
+
+  {
+    // set up the input/output buffers
+    buffer in_buf(in);
+    buffer out_buf(out);
```

Comment:
This example is a little bit misleading because since you are using a buffers you do not really need vectors at the first place.
So it would be interesting to see what is the performance by using just buffers and using some host accessor when you need to initialize or access the data from the host.

Reply:
I agree we don't need vectors AND buffers here, it was just to make it more like our other tutorials.

You don't need vectors, but you still need to copy those buffers from host to device. The fact that we are avoiding ANY copy to/from DDR by using host pointers allows us to avoid that overhead. Using host accessors would just make the memory movement more explicit, no?

Reply:
You can control carefully the memory transfers by picking the right model of accessors and not creating buffers from host data when useless. For example why initializing `out_buf` from `out`? But of course it changes a little bit the code organization.

Reply:
I think there are a couple things here. I will try to break them down.

> why initializing out_buf from out?

So that, when the kernel finishes, the data is automatically gets copied from out_buf to out. The user shouldn't have to explicitly copy to a host pointer or use a host accessor.

> Doesn't that incur a useless copy from out to out_buf before starting the kernel?

Yes. However, we think that is a bug and are trying to fix that in SYCL.

> Can the user not get better performance by explicitly moving memory?

Yes, and that was the point of the explicit_memory_movement tutorial. However, we don't think this should be a users (especially a new users) default. They should start with implicit data movement and change to explicit when they are 1) happy with the functionality and 2) think they get get performance increases.

Lastly, it is worth mentioning, even if the buffers were handled perfectly in this case, the zero-copy data transfer will still perform better since it avoids ANY overhead in copying to/from DDR and does everything over PCIe.